### PR TITLE
Add `UnionArray::into_parts`

### PR DESCRIPTION
**Pull request by mbrobbel**
# Rationale for this change
 
Most arrays have an `into_parts` method to destruct the array into its constituent parts: https://docs.rs/arrow-array/latest/arrow_array/?search=into_parts

# What changes are included in this PR?

This adds that `into_parts` method to `UnionArray`.

# Are there any user-facing changes?

- `arrow_array::array::UnionArray::into_parts`

### CODE REVIEW

File: arrow-array/src/array/union_array.rs [hunk 1]
```diff
@@ -319,6 +319,39 @@ impl UnionArray {
...
+        Vec<Option<ArrayRef>>,
+    ) {
+        let Self {
+            data_type,
+            type_ids,
+            offsets,
+            fields,
+        } = self;
+        (data_type, type_ids, offsets, fields)
+    }
```
Comment (Jefffrey) [inline arrow-array/src/array/union_array.rs:354]: It seems PrimitiveArray is the only other array type that returns a `DataType` as part of its `into_parts()`

Maybe just return the fields, akin to how StructArray does it? (Though unsure about the Sparse/Dense enum)

File: arrow-array/src/array/union_array.rs [hunk 2]
```diff
@@ -319,6 +319,39 @@ impl UnionArray {
...
+    /// # }
+    /// ```
+    #[allow(clippy::type_complexity)]
+    pub fn into_parts(
+        self,
+    ) -> (
+        DataType,
+        ScalarBuffer<i8>,
+        Option<ScalarBuffer<i32>>,
+        Vec<Option<ArrayRef>>,
```
Comment (tustvold) [inline arrow-array/src/array/union_array.rs:345]: This is exposing what is effectively an implementation detail of UnionArray, that it uses a sparse vec to encode type ids. We may want to revisit this in future, e.g. to better handle large type ids

**Comment by tustvold** [comment]
I'm honestly a little torn on this as UnionArray is a bit of an oddity for two reasons:

* Its constructors don't look anything like the other arrays (for historical reasons) and don't match into_parts
* The internal storage of a UnionArray is somewhat an implementation detail w.r.t. how it encodes type ids

I think it would be good to understand the motivating use-case for this, as AFAICT the output of into_parts wouldn't be useful for reconstructing a UnionArray

**Comment by mbrobbel** [comment]
> I think it would be good to understand the motivating use-case for this, as AFAICT the output of into_parts wouldn't be useful for reconstructing a UnionArray

My use-case is [zero-copy conversion](https://github.com/mbrobbel/narrow/blob/2ab7e0436f6530289ebba38c4ce344570a26926f/src/arrow/array/union.rs#L133-L181) to a different union array type ([test example here](https://github.com/mbrobbel/narrow/blob/2ab7e0436f6530289ebba38c4ce344570a26926f/src/arrow/array/union.rs#L221-L244)).

I'm not using it to reconstruct an `arrow_array::array::UnionArray`, however the added test in this PR shows how that can be done.

File: arrow-array/src/array/union_array.rs [hunk 3]
```diff
@@ -1201,4 +1241,70 @@ mod tests {
...
+        assert_eq!(fields.len(), 2);
+
+        let result = UnionArray::try_new(
+            &[0, 1],
+            type_ids.into_inner(),
+            offsets.map(ScalarBuffer::into_inner),
+            field
+                .into_iter()
+                .zip(fields.into_iter().flatten())
+                .collect(),
```
Comment (tustvold) [inline arrow-array/src/array/union_array.rs:1305]: Perhaps `into_parts` could perform this flattening?

Comment (tustvold) [inline arrow-array/src/array/union_array.rs:1305]: I don't think this is correct if the type ids in the signature are out of order

File: arrow-array/src/array/union_array.rs [hunk 4]
```diff
@@ -1201,4 +1241,70 @@ mod tests {
...
+            UnionFields::new(field_type_ids, field.clone())
+        );
+        assert_eq!(union_mode, UnionMode::Dense);
+        assert_eq!(type_ids, [0, 1, 0]);
+        assert!(offsets.is_some());
+        assert_eq!(offsets.as_ref().unwrap(), &[0, 0, 1]);
+        assert_eq!(fields.len(), 2);
+
+        let result = UnionArray::try_new(
+            &[0, 1],
```
Comment (tustvold) [inline arrow-array/src/array/union_array.rs:1270]: Might be worth showing this works with type ids of `[6, 2]` or something

File: arrow-array/src/array/union_array.rs [hunk 5]
```diff
@@ -1201,4 +1241,70 @@ mod tests {
...
+        assert_eq!(union_mode, UnionMode::Sparse);
+        assert_eq!(type_ids, [0, 1, 0]);
+        assert!(offsets.is_none());
+        assert_eq!(fields.len(), 2);
+
+        let result = UnionArray::try_new(
+            &[0, 1],
+            type_ids.into_inner(),
+            offsets.map(ScalarBuffer::into_inner),
+            field
```
Comment (tustvold) [inline arrow-array/src/array/union_array.rs:1302]: I think this flattening should probably be an implementation detail of into_parts

Comment (mbrobbel) [reply arrow-array/src/array/union_array.rs:1302]: Maybe the `fields` argument of `into_parts` could be changed to `impl Iterator<Item = (i8, ArrayRef)>` to hide implementation details?

Comment (tustvold) [reply arrow-array/src/array/union_array.rs:1302]: I think it should probably reconstruct the arrays in the order expected by the constructor

Comment (mbrobbel) [reply arrow-array/src/array/union_array.rs:1302]: Makes sense. The constructor expects the order (of the child arrays) to match the provided fields ids, however, if unknown, one could get these `type_id`s from the `UnionFields` iterator, which has no order guarantee:

https://github.com/apache/arrow-rs/blob/5a0baf1aeeead4c461f1c8d14bfc65c85cfff78e/arrow-schema/src/fields.rs#L422-L427

**Comment by tustvold** [comment]
I took the liberty of removing UnionMode as it is implied by the presence or abscence of the offsets

**Pull request merged**